addons/mail/static/src/models/thread_viewer/thread_viewer_container.js

```diff
@@ -0,0 +1,129 @@
+
+        /**
+         * @private
+         * @returns {string}
+         */
+        _computeStringifiedDomain() {
+            if (this.discuss) {
+                return this.discuss.stringifiedDomain;
+            }
+            return '[]';
+        }
+
+        /**
+         * @private
+         * @returns {mail.thread|undefined}
+         */
+         _computeThread() {
+            if (this.chatter) {
+                return [['link', this.chatter.thread]];
+            }
+            if (this.chatWindow) {
+                return [['link', this.chatWindow.thread]];
+            }
+            if (this.discuss) {
+                return [['link', this.discuss.thread]];
+            }
+            return [['unlink']];
+        }
+
+        /**
+         * @private
+         * @returns {mail.thread_viewer|undefined}
+         */
+        _computeThreadViewer() {
+            if (!this.hasThreadViewer) {
+                return [['unlink']];
+            }
+            if (this.threadViewer) {
+                return [];
+            }
+            return [['create']];
+        }
+
+    }
+
+    ThreadViewerContainer.fields = {
+        /**
+         * Determine whether the current thread viewer should be displayed.
+         *
+         * This field should be controlled by the record which created this
+         * container, and it should otherwise be considered read-only.
+         */
+        hasThreadViewer: attr(),
+        /**
+         * Domain to apply when fetching messages for the current thread.
+         *
+         * This field should be controlled by the record which created this
+         * container, and it should otherwise be considered read-only.
+         */
+        stringifiedDomain: attr({
+            default: '[]',
+        }),
+        /**
+         * Thread currently displayed by this thread viewer.
+         *
+         * This field should be controlled by the record which created this
+         * container, and it should otherwise be considered read-only.
+         */
+        thread: many2one('mail.thread'),
+        /**
+         * List of saved initial scroll positions of thread caches.
+         * Useful to restore scroll position on changing back to this
+         * thread cache. Note that this is only applied when opening
+         * the thread cache, because scroll position may change fast so
+         * save is already throttled.
+         */
+        threadCacheInitialScrollPositions: attr({
+            default: {},
+        }),
+        /**
+         * Thread viewer currently displayed and controlled by this container.
+         */
+        threadViewer: one2one('mail.thread_viewer', {
+            compute: '_computeThreadViewer',
+            dependencies: [
+                'hasThreadViewer',
+            ],
+            inverse: 'container',
+            isCausal: true,
+        }),
+    };
+
+    ThreadViewerContainer.modelName = 'mail.thread_viewer_container';
```

Comment:
I would do the following renames:
- `mail.thread_viewer` => `mail.thread_view`
- `mail.thread_viewer_container` => `mail.thread_viewer`

I feel it's more natural that way. And it gives more direction to desired design, like `mail.discuss` and `mail.chatter` being `mail.thread_viewer`... hinting at `mail.thread_viewer` being a mixin of some sort... `mail.thread_view` is clearer to the intention that this represents an actual view of thread.

Reply:
This change would make the PR totally unreadable and unmaintainable. If that's what we want we must absolutely do a preliminary PR to rename the existing "thread viewer" to "thread view", including file rename that go with it.